src/main/java/io/cryostat/recordings/RecordingMetadataManager.java

```diff
@@ -0,0 +1,202 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.cryostat.recordings;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import io.cryostat.core.log.Logger;
+import io.cryostat.core.sys.FileSystem;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+import org.apache.commons.codec.binary.Base32;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class RecordingMetadataManager {
+
+    public static final String NOTIFICATION_CATEGORY = "RecordingMetadataUpdated";
+
+    private final Path recordingMetadataDir;
+    private final FileSystem fs;
+    private final Gson gson;
+    private final Base32 base32;
+    private final Logger logger;
+
+    private final Map<Pair<String, String>, Map<String, String>> recordingLabelsMap;
+
+    RecordingMetadataManager(
+            Path recordingMetadataDir, FileSystem fs, Gson gson, Base32 base32, Logger logger) {
+        this.recordingMetadataDir = recordingMetadataDir;
+        this.fs = fs;
+        this.gson = gson;
+        this.base32 = base32;
+        this.logger = logger;
+        this.recordingLabelsMap = new ConcurrentHashMap<>();
+    }
+
+    public void load() throws IOException {
+        this.fs.listDirectoryChildren(recordingMetadataDir).stream()
+                .peek(n -> logger.trace("Recording Metadata file: {}", n))
+                .map(recordingMetadataDir::resolve)
+                .map(
+                        path -> {
+                            try {
+                                return fs.readFile(path);
+                            } catch (IOException e) {
+                                logger.warn(e);
+                                return null;
+                            }
+                        })
+                .filter(Objects::nonNull)
+                .map(reader -> gson.fromJson(reader, StoredRecordingMetadata.class))
+                .forEach(
+                        srm ->
+                                recordingLabelsMap.put(
+                                        Pair.of(srm.getTargetId(), srm.getRecordingName()),
+                                        srm.getLabels()));
+    }
+
+    public Future<Map<String, String>> addRecordingLabels(
```

Comment:
Could you add an `addRecordingLabels(String recordingName, Map<String, String> labels)` method definition that just calls through to this one with that magic `ARCHIVES` constant? I think this will clean up the interface a bit, so that calling code doesn't need to know about the "targetId=='archives'" special case so explicitly.

Reply:
Also, I think this method's name is a little misleading - should be it named `setRecordingLabels` instead of `addRecordingLabels`, since calling this will over-write the previous value, rather than joining the old map with the new map?

Reply:
Good point, should be fixed now.